### PR TITLE
SSZ impl fixes

### PR DESCRIPTION
**Pull request by protolambda**
Cleaned up the `minimal_ssz.py` a bit, and fixed some bugs in the progress.
- vectors/lists where mixed up in serialization somewhere, incorrect serialization
- vector <-> list typing unclear at times, resulting in more funny serialization
- improved structure, generalized container encoding a bit using helper functions
- documented the magic

### CODE REVIEW

**Comment by protolambda** [comment]
Tested the ssz implementation with my other SSZ work. I have ~10K generated test-cases (random, zeroed, nil-length, max value, chaotic mix, etc.) generated from this, and ZRNT (after a few bugfixes) produces the same SSZ serialization/tree-roots :tada:

File: test_libs/pyspec/eth2spec/utils/minimal_ssz.py [hunk 1]
```diff
@@ -58,18 +58,39 @@ def __len__(self):
...
+    if typ[:4] == 'uint' and typ[4:] in ['8', '16', '32', '64', '128', '256']:
+        return True
+    # "bool": True or False
+    if typ == 'bool':
+        return True
+    # alias: "byte" -> "uint8"
+    if typ == 'byte':
+        return True
+    # default
+    return False
```
Comment (hwwhww) [inline test_libs/pyspec/eth2spec/utils/minimal_ssz.py]: IMO using `if-elif-else` would be more clear.

File: test_libs/pyspec/eth2spec/utils/minimal_ssz.py [hunk 2]
```diff
@@ -151,42 +198,71 @@ def mix_in_length(root, length):
     return hash(root + length.to_bytes(32, 'little'))
 
 
+# Note: defaults to uint64 for integer type inference due to lack of information.
+# Other integer sizes are still supported, see spec.
```
Comment (hwwhww) [inline test_libs/pyspec/eth2spec/utils/minimal_ssz.py]: This comment could be moved to docstring.

File: test_libs/pyspec/eth2spec/utils/minimal_ssz.py [hunk 3]
```diff
@@ -151,42 +198,71 @@ def mix_in_length(root, length):
...
+            return [infer_type(value[0])]
+        else:
+            # Element type does not matter, list-content size will be encoded regardless, list is empty.
+            return ['uint64']
     elif isinstance(value, (bytes, str)):
         return 'bytes'
     elif isinstance(value, int):
         return 'uint64'
+    elif isinstance(value, int):
+        return 'uint64'
```
Comment (hwwhww) [inline test_libs/pyspec/eth2spec/utils/minimal_ssz.py]: Redundant to the previous `elif isinstance(value, int): ...`

File: test_libs/pyspec/eth2spec/utils/minimal_ssz.py [hunk 4]
```diff
@@ -151,42 +198,71 @@ def mix_in_length(root, length):
...
     else:
         raise Exception("Failed to infer type")
 
-
```
Comment (hwwhww) [inline test_libs/pyspec/eth2spec/utils/minimal_ssz.py]: Let's keep 2 new lines as PEP8!

**Pull request merged**